### DISCUSSION TITLE
Using Github Action without Docker

### DISCUSSION THREAD

**Discussion by rndmh3ro** [ansible-community/ansible-test-gh-action#33] [Q&A] [answered]
I tried using the action to test my collection. My integration tests fail.

Here's the code: https://github.com/T-Systems-MMS/ansible-collection-icinga-director/blob/ansible_2.14/.github/workflows/main.yml#L71

Here's the failing job: https://github.com/T-Systems-MMS/ansible-collection-icinga-director/actions/runs/3096045473/jobs/5011132984

I'm running a service-container on port 80 that is used to test the collection. With the old code (https://github.com/T-Systems-MMS/ansible-collection-icinga-director/blob/master/.github/workflows/main.yml) everything works.

I'm guessing this is because the new action uses docker to run the tests and my former workflow runs the tests natively.

Is there any way to achieve it without docker or am I complete wring here?

**Comment by felixfontein** [comment, marked as answer]
Right now this action only allows running tests with docker (and this is definitely the cause for your problem).

What you could try:
1. Use `pre-test-cmd` to create a `tests/integration/integration_config.yml` file which contains the IP of the main container (or even directly the service container), so that the service can be reached as `<ip_from_that_file>:80`.
2. Adjust your tests to use the variable supplied by `tests/integration/integration_config.yml` to talk to the service.

If you add a default `tests/integration/integration_config.yml` file in your repo that sets the IP to 127.0.0.1 your tests should just continue working for folks who run them locally.

EDIT: The action was updated to tell ansible-test about GitHub's docker network. With this, you are able to access the icinga container as `icinga`. (See the discussion below for details.)

**Reply by rndmh3ro** [reply]
Thanks Felix, for your answer!

It seems harder than I thought. I tried running the following code in the `pre-test-cmd` (https://github.com/T-Systems-MMS/ansible-collection-icinga-director/actions/runs/3097952836/jobs/5015249500) as well as a normal step before the integration step (https://github.com/T-Systems-MMS/ansible-collection-icinga-director/actions/runs/3098023574/jobs/5015407699):

```
             export id=$(docker ps -f 'name=.*icinga.*' -q); echo $id
             export address=$(docker inspect $id -f '{{ .NetworkSettings.IPAddress }}'); echo $address
             sed -i "s/127.0.0.1/${address}/g" tests/integration/integration_config.yml

```

But nothing works.. :(

I'm going to stop working on this for now as I can still use the old actions.

**Reply by webknjaz** [reply]
Yeah, I've been meaning to look into auto-detection of the GitHub's docker network and exposing it via `--docker-network` but haven't gotten to doing that.

Another problem is that ansible-test wasn't designed to handle matrixes in integration tests and this is something it'll need to have implemented for better UX.

Using `--docker` gives us a lot of predictability, but I agree that in special cases, it'd be nice to support not having it.

**Reply by webknjaz** [reply]
> I'm going to stop working on this for now as I can still use the old actions.

I've been suggesting other projects with this issue that they start using this action only for sanity+units and keep integration as is until this is supported.

**Reply by webknjaz** [reply]
Please, check if `ansible-community/ansible-test-gh-action@unstable/v1` fixes your problem accessing the service.

**Reply by rndmh3ro** [reply]
Yes, this works!

I had to change the url from `127.0.0.1` to the service-name `icinga`.

**Reply by webknjaz** [reply]
Great, I'll tag a new release with that soon, then.

By the way, the docs here https://docs.github.com/en/actions/using-containerized-services/about-service-containers#mapping-docker-host-and-service-container-ports suggest that you could use `${{ job.services.icinga.ports[80] }}` to get the exposed port instead of having a static mapping.

**Reply by webknjaz** [reply]
@rndmh3ro v1.9.0 is out at https://github.com/marketplace/actions/ansible-test?version=v1.9.0. Feedback for this release can be posted at https://github.com/ansible-community/ansible-test-gh-action/discussions/37. Please, make sure to update your pin from the unstable branch to a tag or a commit sha.

@felixfontein could you update your answer to mention this?
